jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -365,8 +365,10 @@ protected boolean passConditionalHeaders(Request request, Response response, Htt
                         String matched = matchesEtag(etag, ifnm);
                         if (matched != null)
                         {
+                            // If we do sendError it clears the ETag header.
+                            response.setStatus(HttpStatus.NOT_MODIFIED_304);
                             response.getHeaders().put(HttpHeader.ETAG, matched);
-                            writeHttpError(request, response, callback, HttpStatus.NOT_MODIFIED_304);
+                            response.write(true, null, callback);
```

Comment:
This needs to be handed to `writeHttpError` so that the DefaultServlet can override and use the Servlet `HttpServletResponse.sendError()` API.

Otherwise application provided filters, security, wrappers, etc are not executed.